nodejs/awsx/apigateway/apikey.ts

```diff
@@ -0,0 +1,147 @@
+ */
+export function createAssociatedAPIKeys(name: string, args: APIKeyArgs): AssociatedAPIKeys {
+    const usagePlan = getUsagePlan(name, args);
+    const keys = getKeys(name, args, usagePlan);
+
+    return {
+        usagePlan: usagePlan,
+        keys: keys,
+    };
+}
+
+/** @internal */
+function getUsagePlan(name: string, args: APIKeyArgs): aws.apigateway.UsagePlan {
+    if (args.usagePlan && args.usagePlan instanceof aws.apigateway.UsagePlan) {
+        if (args.apis) {
+            throw new Error("cannot define both [args.apis] and an existing usagePlan [args.usagePlan]");
+        }
+        return args.usagePlan;
+    } else {
+        let usagePlanArgs: aws.apigateway.UsagePlanArgs | undefined = args.usagePlan;
+
+        if (args.apis) {
+            const stages = [];
+
+            for (const a of args.apis) {
+                stages.push({
+                    apiId: a.restAPI.id,
+                    stage: a.stage.stageName,
+                });
+            }
+
+            if (args.usagePlan) {
+                if (args.usagePlan.apiStages) {
+                    throw new Error("cannot define both [args.apis] and [args.usagePlan.apiStages]");
+                }
+
+                usagePlanArgs = {
+                    apiStages: stages,
+                    description: args.usagePlan.description,
+                    name: args.usagePlan.name,
+                    productCode: args.usagePlan.productCode,
+                    quotaSettings: args.usagePlan.quotaSettings,
+                    throttleSettings: args.usagePlan.throttleSettings,
+                };
+            } else {
+                usagePlanArgs = {
+                    apiStages: stages,
+                };
+            }
+        }
+        return new aws.apigateway.UsagePlan(name, usagePlanArgs);
+    }
+}
+
+/** @internal */
+function getKeys(name: string, args: APIKeyArgs, usagePlan: aws.apigateway.UsagePlan): Key[] {
+    const keys: Key[] = [];
+
+    if (args.apiKeys) {
+        for (let i = 0; i < args.apiKeys.length; i++) {
+            const currKey = args.apiKeys[i];
+            let apikey: aws.apigateway.ApiKey;
+
+            if (isAPIKey(currKey)) {
+                apikey = currKey;
+            } else {
+                apikey = new aws.apigateway.ApiKey(name + "-apikey-" + i, currKey);
+            }
+
+            const usagePlanKey = new aws.apigateway.UsagePlanKey(name + "-usage-plan-key-" + i, {
+                keyId: apikey.id,
+                keyType: "API_KEY",
+                usagePlanId: usagePlan.id,
+            });
+
+            keys.push({
+                apikey: apikey,
+                usagePlanKey: usagePlanKey,
+            });
+        }
+    }
+    return keys;
+}
+
+/** @internal */
+export function isAPIKey(apikey: aws.apigateway.ApiKey | aws.apigateway.ApiKeyArgs): apikey is aws.apigateway.ApiKey {
+    return (<aws.apigateway.ApiKey>apikey).value !== undefined;
```

Comment:
sounds ike this should be CustomResult.isInstance?

Reply:
I think most of the problems arose when we had a Pulumi.Input<T> and a CustomResource.. but I will change these all to be safe :)